cpp/src/lance/io/exec/project.cc

```diff
@@ -70,14 +72,21 @@ ::arrow::Result<std::unique_ptr<Project>> Project::Make(
       ARROW_ASSIGN_OR_RAISE(child, Limit::Make(limit.value(), offset, std::move(child)));
     }
   }
-  return std::unique_ptr<Project>(new Project(std::move(child)));
+  return std::unique_ptr<Project>(new Project(std::move(child), std::move(projected_schema)));
 }
 
 std::string Project::ToString() const { return "Project"; }
 
 ::arrow::Result<ScanBatch> Project::Next() {
+
   assert(child_);
-  return child_->Next();
+  ARROW_ASSIGN_OR_RAISE(auto batch, child_->Next());
+  if (batch.eof()) {
+    return batch;
+  }
+  ARROW_ASSIGN_OR_RAISE(auto projected_batch,
+                        lance::arrow::ApplyProjection(batch.batch, *projected_schema_));
+  return ScanBatch(projected_batch, batch.batch_id, batch.indices);
```

Comment:
Maybe makes sense to add ScanBatch.Project ?

Reply:
Sure, I can move this to ScanBatch.